### PR TITLE
Timezone in App

### LINKED ISSUE #2558: Timezone Middleware with Apollo
**Is your feature request related to a problem? Please describe.**
Implement a middleware which converts the time depending on user's app.

**Describe the solution you'd like**
1) For all the requests, we convert all UTC time from DB to client.
2) For all the mutations, we convert all local to UTC time via middleware.

**Describe alternatives you've considered**
Saving and changing all the components depending on timezone.

**Additional context**
Reference: https://github.com/PalisadoesFoundation/talawa-admin/pull/2234

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by pranshugupta54**
**What kind of change does this PR introduce?**

Feature: timezone

**Issue Number:**

Fixes #2558 

**Did you add tests for your changes?**

yes

**Summary**

Solution: Instead of making significant changes to the Mobile/API repositories, I modified how the GraphQL API interacts with Apollo. Now, we're using middleware to handle time conversions: function converts UTC to the local time zone during queries, and function converts local time back to UTC during mutations.

Alternative Approach: I initially spent a lot of time on a different approach that we had discussed and planned earlier according to our issue and project planning. We had planned to change the database by introducing a timezone field and then saving data based on the organization. However, since we don't intend to give users the option to switch timezones like Google Calendar, we can directly modify the API requests and responses to implement this.

**Does this PR introduce a breaking change?**

NO

**Other information**

NA

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa/blob/master/CONTRIBUTING.md)?**

Yes

## Release Notes

- **New Features**
	- Introduced utility functions for comprehensive date and time conversions, enhancing data handling across the application.
	- Enhanced date conversion logic in GraphQL operations for better date management.
	- Improved event creation time formatting for accurate representation.

- **Bug Fixes**
	- Enhanced error handling in volunteer group fetching to manage exceptions more effectively.

- **Tests**
	- Added a new test suite for time conversion utilities to validate functionality and ensure reliability.
	- Expanded test coverage for the EventService with new tests for volunteer group management, including creation, removal, and updates.

### CODE REVIEW

**Comment by palisadoes** [comment]
@pranshugupta54 

Please try to get the code coverage up so that the tests pass.

**Comment by palisadoes** [comment]
1. Does the mobile client use the local timezone when doing it's datetime calculations?
2. Is this true for the changes you made to talawa-admin?

**Comment by Dante291** [comment]
> @palisadoes
> 
> > Please try to get the code coverage up so that the tests pass.
> 
> The coverage is low mainly due to changes in `[lib/services/event_service.dart]` file which was actually a part of different issue. @Dante291 helped me fix it as this issue was blocking me so he'll add the tests for it in his PR.
> 
> I will be adding lines ignore for the some lines in `[lib/services/database_mutation_functions.dart]` because we won't be doing timezone tests for mock as we discussed earlier. (Mocks are static data so if we add tests for it then we'll need different test-mocks for each and every timezone)
> 
> > Does the mobile client use the local timezone when doing it's datetime calculations?
> > Is this true for the changes you made to talawa-admin?
> 
> Yes, it's similar to Talawa-Admin. Now it's a middleware to change the time. I've tested this and now whenever I make a new event using IST time then it converts and saves in UTC in database, and at the time of viewing any event - it automatically converts back to IST (my local time) from UTC. So, for a user it's always shown in local time for them, but we store in UTC in database.

Just dont make these changes in your PR, I will do that in mine and add tests too

**Comment by Dante291** [comment]
also please attach a clip of app working properly after making these changes to ensure proper working

**Comment by palisadoes** [comment]
@noman2002 PTAL.

The aim is for the mobile app to interact with the API only using UTC times. This will help with converting calendar entries to the correct local time when the end user could be in a different timezone from the client.

Work in Admin and API has already been completed

File: lib/services/database_mutation_functions.dart [hunk 1]
```diff
@@ -209,6 +227,13 @@ class DataBaseMutationFunctions {
             result.exception!,
           );
         } else if (result.data != null && result.isConcrete) {
+          // coverage:ignore-start
+          traverseAndConvertDates(
+            result.data ?? <String, dynamic>{},
+            convertUTCToLocal,
+            splitDateTimeLocal,
+          );
+          // coverage:ignore-end
```
Comment (noman2002) [inline lib/services/database_mutation_functions.dart:236]: @pranshugupta54 Why we need this in mutation ?

Comment (noman2002) [reply lib/services/database_mutation_functions.dart:236]: Got it, so we are storing it in UTC and displaying in local, correct ?

**Review by noman2002 — APPROVED** [review]
LGTM, good work @pranshugupta54 .

**Comment by Dante291** [comment]
@palisadoes @pranshugupta54  This PR breaks the app, no server call are successful after this got merged. need to revert it.

**Comment by Dante291** [comment]
> @Dante291, did you try making the changes in that event file. This PR won't work without those changes.

Yup I did that, also those changes were related to event service only so even if anything breaks if some error is occurring due to that, it should be only related to events, but after this got merged neither user data is loading now the posts, basically nothing.

**Comment by palisadoes** [comment]
It's merged

**Pull request merged**